hazelcast/src/main/java/com/hazelcast/sql/impl/expression/predicate/ComparisonPredicate.java

```diff
@@ -126,6 +147,47 @@ public Boolean eval(Row row, ExpressionEvalContext context) {
         }
     }
 
+    private boolean compareRows(RowValue left, RowValue right) {
+        if (left.getValues().size() != right.getValues().size()) {
+            return false;
+        }
+
+        for (int i = 0; i < left.getValues().size(); i++) {
+            final Object leftVal = left.getValues().get(i);
+            final Object rightVal = right.getValues().get(i);
+
+            if (Objects.equals(leftVal, rightVal)) {
+                continue;
+            }
+
+            if (leftVal == null || rightVal == null) {
+                return false;
+            }
+
+            if (leftVal instanceof RowValue && rightVal instanceof RowValue) {
+                if (!compareRows((RowValue) leftVal, (RowValue) rightVal)) {
+                    return false;
+                } else {
+                    continue;
+                }
+            }
+
+            final Converter leftConverter = Converters.getConverter(leftVal.getClass());
```

Comment:
Type coercion should be done at the validation phase, here we should need no conversion.